src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs

```diff
@@ -119,6 +121,503 @@ internal int EffectiveMaxConcurrencyLevel
+        private readonly Action<TValue, ParallelLoopState> _bodyWithState;
+
+        public WorkerWithStateFactory(Action<TValue, ParallelLoopState> bodyWithState)
+        {
+            _bodyWithState = bodyWithState;
+        }
+
+        public IWorkerBody<TValue> CreateWorkerBody(ParallelLoopStateFlags sharedPStateFlags)
+            => new WorkerWithState<TValue, TIndex>(sharedPStateFlags, _bodyWithState);
+    }
+
+    internal sealed class WorkerWithStateAndIndexFactory<TValue, TIndex> : IWorkerBodyFactory<TValue> where TIndex: INumber<TIndex>
+    {
+        private readonly Action<TValue, ParallelLoopState, TIndex> _bodyWithStateAndIndex;
+
+        public WorkerWithStateAndIndexFactory(Action<TValue, ParallelLoopState, TIndex> bodyWithStateAndIndex)
+        {
+            _bodyWithStateAndIndex = bodyWithStateAndIndex;
+        }
+
+        public IWorkerBody<TValue> CreateWorkerBody(ParallelLoopStateFlags sharedPStateFlags)
+            => new WorkerWithStateAndIndex<TValue, TIndex>(sharedPStateFlags, _bodyWithStateAndIndex);
+    }
+
+
+    internal sealed class WorkerWithLocalFactory<TValue, TIndex, TLocal> : IWorkerBodyFactory<TValue> where TIndex: INumber<TIndex>
+    {
+        private readonly Func<TValue, ParallelLoopState, TLocal, TLocal> _bodyWithLocal;
+        private readonly Func<TLocal> _localInit;
+        private readonly Action<TLocal>? _localFinally;
+
+        public WorkerWithLocalFactory(Func<TValue, ParallelLoopState, TLocal, TLocal> bodyWithLocal, Func<TLocal> localInit, Action<TLocal>? localFinally)
+        {
+            _bodyWithLocal = bodyWithLocal;
+            _localInit = localInit;
+            _localFinally = localFinally;
+        }
+
+        public IWorkerBody<TValue> CreateWorkerBody(ParallelLoopStateFlags sharedPStateFlags)
+            => new WorkerWithStateAndLocal<TValue, TIndex, TLocal>(sharedPStateFlags, _bodyWithLocal, _localInit, _localFinally);
+    }
+
+    internal sealed class WorkerWithEveryThingFactory<TValue, TIndex, TLocal> : IWorkerBodyFactory<TValue> where TIndex: INumber<TIndex>
+    {
+        private readonly Func<TValue, ParallelLoopState, TIndex, TLocal, TLocal> _bodyWithLocal;
+        private readonly Func<TLocal> _localInit;
+        private readonly Action<TLocal>? _localFinally;
+
+        public WorkerWithEveryThingFactory(Func<TValue, ParallelLoopState, TIndex, TLocal, TLocal> bodyWithLocal, Func<TLocal> localInit, Action<TLocal>? localFinally)
+        {
+            _bodyWithLocal = bodyWithLocal;
+            _localInit = localInit;
+            _localFinally = localFinally;
+        }
+
+        public IWorkerBody<TValue> CreateWorkerBody(ParallelLoopStateFlags sharedPStateFlags)
+            => new WorkerWithEverything<TValue, TIndex, TLocal>(sharedPStateFlags, _bodyWithLocal, _localInit, _localFinally);
+    }
+
+    #endregion
+
+    #region Workers
+
+    internal interface IWorkerFactory<in TSource, TInput, TValue, out TIndex> where TIndex: INumber<TIndex>
```

Comment:
This interface started as a factory only interface but now contains more loop run related metadata.
Is there a better name for this?